### PR TITLE
Proposal: Introduce DetourUpdateAllOtherThreads

### PR DESCRIPTION
**Pull request by adams85**
Introduce a new public function which updates all threads apart from the current one in a single operation.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/Detours/pull/233)

### CODE REVIEW

File: src/detours.cpp [hunk 1]
```diff
@@ -1524,6 +1524,24 @@ struct DetourThread
 {
     DetourThread *      pNext;
     HANDLE              hThread;
+    BOOL                fCloseThreadHandleOnDestroy;
+
+    DetourThread() {
```
Comment (sylveon) [inline src/detours.cpp:1529]: Allman braces

File: src/detours.cpp [hunk 2]
```diff
@@ -1524,6 +1524,24 @@ struct DetourThread
 {
     DetourThread *      pNext;
     HANDLE              hThread;
+    BOOL                fCloseThreadHandleOnDestroy;
```
Comment (sylveon) [inline src/detours.cpp:1527]: If we're giving this class ownership semantics, delete the copy constructor and assignment operator.

File: src/detours.cpp [hunk 3]
```diff
@@ -1980,12 +2003,57 @@ LONG WINAPI DetourUpdateThread(_In_ HANDLE hThread)
...
+    _In_ HANDLE ThreadHandle,
+    _In_ ACCESS_MASK DesiredAccess,
+    _In_ ULONG HandleAttributes,
+    _In_ ULONG Flags,
+    _Out_ PHANDLE NewThreadHandle
+);
+
+LONG WINAPI DetourUpdateAllOtherThreads()
+{
+    _NtGetNextThread NtGetNextThread = (_NtGetNextThread)GetProcAddress(GetModuleHandle(TEXT("ntdll.dll")), "NtGetNextThread");
```
Comment (sylveon) [inline src/detours.cpp:2030]: Try avoiding undocumented APIs if possible: Windows 8.1 and up have [PssCaptureSnapshot](https://docs.microsoft.com/en-us/windows/win32/api/processsnapshot/nf-processsnapshot-psscapturesnapshot) that you can dynamically query for. Using NtGetNextThread should be fine on older platforms.

File: src/detours.cpp [hunk 4]
```diff
@@ -1980,12 +2003,57 @@ LONG WINAPI DetourUpdateThread(_In_ HANDLE hThread)
...
+        if (!NT_SUCCESS(status)) {
+            if (status != STATUS_NO_MORE_ENTRIES) {
+                DETOUR_TRACE("Failed to enumerate process threads.\r\n");
+                return ERROR_FUNCTION_FAILED;
+            }
+
+            return NO_ERROR;
+        }
+
+        if (currentThreadId != GetThreadId(hThread)) {
```
Comment (sylveon) [inline src/detours.cpp:2051]: GetThreadId was introduced in Vista but Detours supports XP. This won't compile. See #80.

**Comment by adams85** [comment]
I've addressed the first two problems. The other two (regarding the Windows APIs used) would take more time than I can spare now...

**Comment by sonyps5201314** [comment]
I have already introduced the DetourUpdateAllOtherThreads API in [this PR](https://github.com/microsoft/Detours/pull/144) a long time ago. Your code seems to be a version that modified me a long time ago. The NtGetNextThread and GetThreadId you use do not exist in XP, so your code is compatible Question, does not apply to all systems supported by detours. And my code has been repeatedly revised and tested extensively, so it is stable.

**Comment by sonyps5201314** [comment]
@[adams85](https://github.com/adams85)
Thanks very much!

**Pull request closed without merging**